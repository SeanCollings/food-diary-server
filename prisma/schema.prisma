// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

// https://www.prisma.io/nestjs
// https://docs.nestjs.com/recipes/prisma
// https://blog.logrocket.com/how-to-use-nestjs-prisma/

generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model User {
  id        Int        @id @default(autoincrement())
  name      String
  email     String     @unique
  password  String
  avatar    String?
  createdAt DateTime   @default(now())
  lastLogin DateTime   @default(now())
  shareLink ShareLink?
  diaryDay  DiaryDay[]

  userPreferenceShowDayStreak       Boolean  @default(true)
  userPreferenceShowWeeklyExcercise Boolean?
  userPreferenceShowWeeklyWater     Boolean?

  statLastActivity DateTime?
  statDayStreak    Int?
}

model ShareLink {
  id       Int     @id @default(autoincrement())
  user     User    @relation(fields: [userId], references: [id])
  userId   Int     @unique
  link     String  @unique
  isShared Boolean
}

model DiaryDay {
  id                       String  @id @default(uuid())
  user                     User    @relation(fields: [userId], references: [id])
  userId                   Int
  date                     String
  mealBreakfast            Json?
  mealSnack1               Json?
  mealLunch                Json?
  mealSnack2               Json?
  mealDinner               Json?
  wellnessWater            Float?
  wellnessTeaCoffee        Float?
  wellnessAlcohol          Float?
  wellnessExcercise        String?
  wellnessExcerciseDetails String?

  hasMealBreakfast     Boolean?
  hasMealSnack1        Boolean?
  hasMealLunch         Boolean?
  hasMealSnack2        Boolean?
  hasMealDinner        Boolean?
  hasWellnessExcercise Boolean?

  @@unique([userId, date])
}
