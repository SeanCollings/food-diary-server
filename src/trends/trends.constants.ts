export const WEEK_TOTAL_VALUES = 7;
export const TRENDS_WEEK_LEGEND = ['S', 'M', 'T', 'W', 'T', 'F', 'S'];
