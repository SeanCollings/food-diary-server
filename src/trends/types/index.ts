import { IsIn, IsNotEmpty, IsString } from 'class-validator';
import { WellnessTypesEnum } from '@/wellness/types';

export interface RequestWithUser extends Request {
  user: { userId: number; email: string };
}

export type TrendType = 'week' | 'month';

export class GetTrendsQuery {
  @IsString()
  @IsNotEmpty()
  @IsIn(['week', 'month'])
  type: TrendType;
}

export type BeverageTypes = Exclude<
  WellnessTypesEnum,
  WellnessTypesEnum.EXCERCISE
>;
