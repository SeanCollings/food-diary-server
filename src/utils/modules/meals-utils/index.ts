export * from './column-lookup';
export * from './build-meal-content';
export * from './remove-meal-content';
export * from './update-meal-content';
export * from './get-meal-contents';
