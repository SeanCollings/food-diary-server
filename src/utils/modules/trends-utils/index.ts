export * from './get-all-dates-for-trend-type';
export * from './get-meal-trend-data';
export * from './get-beverage-trend-data';
export * from './get-exercise-trend-data';
