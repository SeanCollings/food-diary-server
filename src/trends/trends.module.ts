import { PrismaService } from '@/prisma.service';
import { Module } from '@nestjs/common';
import { TrendsController } from './trends.controller';
import { TrendsService } from './trends.service';

@Module({
  controllers: [TrendsController],
  providers: [TrendsService, PrismaService],
})
export class TrendsModule {}
