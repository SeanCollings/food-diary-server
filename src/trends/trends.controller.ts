import { JwtAuthGuard } from '@/auth/jwt-auth.guard';
import { Controller, Get, Query, Request, UseGuards } from '@nestjs/common';
import { TrendsService } from './trends.service';
import { GetTrendsQuery, RequestWithUser } from './types';

@Controller('trends')
export class TrendsController {
  constructor(private mealsService: TrendsService) {}

  @Get('/meal-trends')
  @UseGuards(JwtAuthGuard)
  async getMealTrends(
    @Request() req: RequestWithUser,
    @Query() query: GetTrendsQuery,
  ) {
    return this.mealsService.getMealTrends(req.user.userId, query.type);
  }

  @Get('/beverage-trends')
  @UseGuards(JwtAuthGuard)
  async getBeverageTrends(
    @Request() req: RequestWithUser,
    @Query() query: GetTrendsQuery,
  ) {
    return this.mealsService.getBeverageTrends(req.user.userId, query.type);
  }

  @Get('/excercise-trends')
  @UseGuards(JwtAuthGuard)
  async getExcerciseTrends(
    @Request() req: RequestWithUser,
    @Query() query: GetTrendsQuery,
  ) {
    return this.mealsService.getExcerciseTrends(req.user.userId, query.type);
  }
}
