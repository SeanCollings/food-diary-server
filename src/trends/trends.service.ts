import { EMealType } from '@/lib/interfaces/meals';
import { PrismaService } from '@/prisma.service';
import {
  formatToServerDate,
  getDateRangeBackTillDayOfWeek,
  getMidnightISODaysInMonth,
  getMonthAndYearFromDate,
  sortDateArray,
} from '@/utils/date-utils';
import { convertTimeStringToMinutes } from '@/utils/time-utils';
import { Injectable } from '@nestjs/common';
import { DiaryDay } from '@prisma/client';
import { TRENDS_WEEK_LEGEND, WEEK_TOTAL_VALUES } from './trends.constants';
import { BeverageTypes, TrendType } from './types';

const getAllDatesForType = (type: TrendType) => {
  const today = formatToServerDate(new Date());
  let dateRangeFromToday: string[];

  if (type === 'week') {
    dateRangeFromToday = getDateRangeBackTillDayOfWeek(today, 1);
  } else {
    const [month, year] = getMonthAndYearFromDate();
    dateRangeFromToday = getMidnightISODaysInMonth(month, year);
  }

  const sortedDates = sortDateArray(dateRangeFromToday, 'asc');

  return sortedDates;
};

const getMealTrendData = (
  type: TrendType,
  allDates: string[],
  diaryDays: DiaryDay[],
) => {
  const mealTotals: { [key in EMealType]: number } = {
    breakfast: 0,
    snack_1: 0,
    lunch: 0,
    snack_2: 0,
    dinner: 0,
  };

  const legend =
    type === 'week'
      ? TRENDS_WEEK_LEGEND
      : allDates.map((_, index) => `${index + 1}`);

  const mealsPerDay = allDates.reduce((acc, date, index) => {
    const entry = diaryDays.find((day) => day.date === date);

    if (entry) {
      const breakfast = entry.hasMealBreakfast ? 1 : 0;
      const snack1 = entry.hasMealSnack1 ? 1 : 0;
      const lunch = entry.hasMealLunch ? 1 : 0;
      const snack2 = entry.hasMealSnack2 ? 1 : 0;
      const dinner = entry.hasMealDinner ? 1 : 0;

      mealTotals.breakfast += breakfast;
      mealTotals.snack_1 += snack1;
      mealTotals.lunch += lunch;
      mealTotals.snack_2 += snack2;
      mealTotals.dinner += dinner;

      acc.push({
        id: `day_${index + 1}`,
        meals: [breakfast, snack1, lunch, snack2, dinner],
      });
    } else {
      acc.push({
        id: `day_${index + 1}`,
        meals: [0, 0, 0, 0, 0],
      });
    }
    return acc;
  }, [] as { id: string; meals: number[] }[]);

  return { legend, mealTotals, mealsPerDay };
};
const getBeverageTrendData = (
  type: TrendType,
  allDates: string[],
  diaryDays: DiaryDay[],
) => {
  let highestValue = 0;
  const defaultDay: {
    [key in BeverageTypes]: number;
  } = {
    water: 0,
    tea_coffee: 0,
    alcohol: 0,
  };

  const legend =
    type === 'week'
      ? TRENDS_WEEK_LEGEND
      : allDates.map((_, index) => `${index + 1}`);

  const beveragesPerDay = allDates.reduce((acc, date) => {
    const entry = diaryDays.find((day) => day.date === date);

    if (entry) {
      const water = entry.wellnessWater ?? 0;
      const tea_coffee = entry.wellnessTeaCoffee ?? 0;
      const alcohol = entry.wellnessAlcohol ?? 0;
      const highestValueInEntry = Math.max(...[water, tea_coffee, alcohol]);

      if (highestValueInEntry > highestValue) {
        highestValue = highestValueInEntry;
      }

      acc.push({
        water,
        tea_coffee,
        alcohol,
      });
    } else {
      acc.push(defaultDay);
    }

    return acc;
  }, [] as { [key in BeverageTypes]: number }[]);

  return { beveragesPerDay, highestValue, legend };
};
const getExceriseTrendData = (
  type: TrendType,
  allDates: string[],
  diaryDays: DiaryDay[],
) => {
  let highestValue = 0;

  const legend =
    type === 'week'
      ? TRENDS_WEEK_LEGEND
      : allDates.map((_, index) => `${index + 1}`);

  const excercisePerDay = allDates.reduce((acc, date) => {
    const entry = diaryDays.find((day) => day.date === date);

    if (entry) {
      const excerciseTime = entry.wellnessExcercise || '';
      const time = convertTimeStringToMinutes(excerciseTime);

      if (time > highestValue) {
        highestValue = time;
      }

      acc.push(time);
    } else {
      acc.push(0);
    }

    return acc;
  }, [] as number[]);

  return { excercisePerDay, highestValue, legend };
};

@Injectable()
export class TrendsService {
  constructor(private prisma: PrismaService) {}

  async getMealTrends(userId: number, type: TrendType) {
    const allDates = getAllDatesForType(type);

    const diaryDays = await this.prisma.diaryDay.findMany({
      where: {
        userId,
        date: { in: allDates },
      },
    });

    const trendData = getMealTrendData(type, allDates, diaryDays);
    return {
      totalValues: type === 'week' ? WEEK_TOTAL_VALUES : allDates.length,
      ...trendData,
    };
  }

  async getBeverageTrends(userId: number, type: TrendType) {
    const allDates = getAllDatesForType(type);

    const diaryDays = await this.prisma.diaryDay.findMany({
      where: {
        userId,
        date: { in: allDates },
      },
    });

    const trendData = getBeverageTrendData(type, allDates, diaryDays);
    return trendData;
  }

  async getExcerciseTrends(userId: number, type: TrendType) {
    const allDates = getAllDatesForType(type);

    const diaryDays = await this.prisma.diaryDay.findMany({
      where: {
        userId,
        date: { in: allDates },
      },
    });

    const trendData = getExceriseTrendData(type, allDates, diaryDays);
    return trendData;
  }
}
