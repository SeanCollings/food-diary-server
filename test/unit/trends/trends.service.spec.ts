import { Test, TestingModule } from '@nestjs/testing';
import { DeepMockProxy, mockDeep } from 'jest-mock-extended';
import { DiaryDay, PrismaClient } from '@prisma/client';
import { TrendsService } from '@/trends/trends.service';
import { PrismaService } from '@/prisma.service';
import { getAllDatesForType } from '@/utils/modules/trends-utils';

jest.mock('@utils/modules/trends-utils', () => ({
  ...jest.requireActual('@utils/modules/trends-utils'),
  getAllDatesForType: jest.fn(),
}));

describe('TrendsService', () => {
  const mockUserId = 1234;
  const mockWeekDates = [
    '2023-04-25T22:00:00.000Z',
    '2023-04-26T22:00:00.000Z',
    '2023-04-27T22:00:00.000Z',
  ];
  const mockMonthDates = [
    '2023-03-31T22:00:00.000Z',
    '2023-04-01T22:00:00.000Z',
    '2023-04-02T22:00:00.000Z',
    '2023-04-03T22:00:00.000Z',
    '2023-04-04T22:00:00.000Z',
  ];
  const mockDiaryDaysWeek: Partial<DiaryDay>[] = [
    {
      id: '1',
      userId: mockUserId,
      date: '2023-04-27T22:00:00.000Z',
      hasMealBreakfast: true,
      hasMealSnack1: true,
      hasMealDinner: true,
      wellnessWater: 2,
      wellnessAlcohol: 1,
      wellnessExcercise: '01:15',
      wellnessExcerciseDetails: 'mock exercise details',
    },
    {
      id: '2',
      userId: mockUserId,
      date: '2023-04-26T22:00:00.000Z',
      hasMealBreakfast: true,
      hasMealLunch: true,
      wellnessWater: 4,
      wellnessTeaCoffee: 2,
    },
  ];
  const mockDiaryDaysMonth: Partial<DiaryDay>[] = [
    {
      id: '1',
      userId: mockUserId,
      date: '2023-03-31T22:00:00.000Z',
      hasMealBreakfast: true,
      hasMealSnack1: true,
      hasMealLunch: true,
      hasMealSnack2: true,
      hasMealDinner: true,
      wellnessWater: 2,
      wellnessAlcohol: 1,
      wellnessExcercise: '01:15',
      wellnessExcerciseDetails: 'mock exercise details',
    },
    {
      id: '2',
      userId: mockUserId,
      date: '2023-04-01T22:00:00.000Z',
      hasMealBreakfast: true,
      hasMealLunch: true,
      wellnessWater: 4,
    },
    {
      id: '3',
      userId: mockUserId,
      date: '2023-04-03T22:00:00.000Z',
      hasMealBreakfast: true,
      hasMealSnack1: true,
      hasMealLunch: true,
      hasMealDinner: true,
      wellnessWater: 2,
      wellnessAlcohol: 0,
      wellnessExcercise: '00:30',
      wellnessExcerciseDetails: 'other mock excercise details',
    },
    {
      id: '4',
      userId: mockUserId,
      date: '2023-04-04T22:00:00.000Z',
      hasMealBreakfast: true,
      hasMealLunch: true,
      hasMealDinner: true,
      wellnessTeaCoffee: 3,
    },
  ];

  let service: TrendsService;
  let prisma: DeepMockProxy<PrismaClient>;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [TrendsService, PrismaService],
    })
      .overrideProvider(PrismaService)
      .useValue(mockDeep<PrismaClient>())
      .compile();

    service = module.get<TrendsService>(TrendsService);
    prisma = module.get(PrismaService);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  describe('getMealTrends', () => {
    it('should get meal trends for a week', async () => {
      (getAllDatesForType as jest.Mock).mockReturnValue(mockWeekDates);

      prisma.diaryDay.findMany.mockResolvedValue(mockDiaryDaysWeek as any);
      const result = await service.getMealTrends(mockUserId, 'week');
      expect(result).toMatchSnapshot();
    });

    it('should get meal trends for a month', async () => {
      (getAllDatesForType as jest.Mock).mockReturnValue(mockMonthDates);
      prisma.diaryDay.findMany.mockResolvedValue(mockDiaryDaysMonth as any);

      const result = await service.getMealTrends(mockUserId, 'month');
      expect(result).toMatchSnapshot();
    });
  });

  describe('getBeverageTrends', () => {
    it('should get beverage trends for a week', async () => {
      (getAllDatesForType as jest.Mock).mockReturnValue(mockWeekDates);
      prisma.diaryDay.findMany.mockResolvedValue(mockDiaryDaysWeek as any);

      const result = await service.getBeverageTrends(mockUserId, 'week');
      expect(result).toMatchSnapshot();
    });

    it('should get beverage trends for a month', async () => {
      (getAllDatesForType as jest.Mock).mockReturnValue(mockMonthDates);
      prisma.diaryDay.findMany.mockResolvedValue(mockDiaryDaysMonth as any);

      const result = await service.getBeverageTrends(mockUserId, 'month');
      expect(result).toMatchSnapshot();
    });
  });

  describe('getExcerciseTrends', () => {
    it('should get excercise trends for a week', async () => {
      (getAllDatesForType as jest.Mock).mockReturnValue(mockWeekDates);
      prisma.diaryDay.findMany.mockResolvedValue(mockDiaryDaysWeek as any);

      const result = await service.getExcerciseTrends(mockUserId, 'week');
      expect(result).toMatchSnapshot();
    });

    it('should get excercise trends for a month', async () => {
      (getAllDatesForType as jest.Mock).mockReturnValue(mockMonthDates);
      prisma.diaryDay.findMany.mockResolvedValue(mockDiaryDaysMonth as any);

      const result = await service.getExcerciseTrends(mockUserId, 'month');
      expect(result).toMatchSnapshot();
    });
  });
});
