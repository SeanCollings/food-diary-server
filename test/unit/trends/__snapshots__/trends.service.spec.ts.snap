// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrendsService getBeverageTrends should get beverage trends for a month 1`] = `
{
  "beveragesPerDay": [
    {
      "alcohol": 1,
      "tea_coffee": 0,
      "water": 2,
    },
    {
      "alcohol": 0,
      "tea_coffee": 0,
      "water": 4,
    },
    {
      "alcohol": 0,
      "tea_coffee": 0,
      "water": 0,
    },
    {
      "alcohol": 0,
      "tea_coffee": 0,
      "water": 2,
    },
    {
      "alcohol": 0,
      "tea_coffee": 3,
      "water": 0,
    },
  ],
  "highestValue": 4,
  "legend": [
    "1",
    "2",
    "3",
    "4",
    "5",
  ],
}
`;

exports[`TrendsService getBeverageTrends should get beverage trends for a week 1`] = `
{
  "beveragesPerDay": [
    {
      "alcohol": 0,
      "tea_coffee": 0,
      "water": 0,
    },
    {
      "alcohol": 0,
      "tea_coffee": 2,
      "water": 4,
    },
    {
      "alcohol": 1,
      "tea_coffee": 0,
      "water": 2,
    },
  ],
  "highestValue": 4,
  "legend": [
    "M",
    "T",
    "W",
    "T",
    "F",
    "S",
    "S",
  ],
}
`;

exports[`TrendsService getExcerciseTrends should get excercise trends for a month 1`] = `
{
  "excercisePerDay": [
    75,
    0,
    0,
    30,
    0,
  ],
  "highestValue": 75,
  "legend": [
    "1",
    "2",
    "3",
    "4",
    "5",
  ],
}
`;

exports[`TrendsService getExcerciseTrends should get excercise trends for a week 1`] = `
{
  "excercisePerDay": [
    0,
    0,
    75,
  ],
  "highestValue": 75,
  "legend": [
    "M",
    "T",
    "W",
    "T",
    "F",
    "S",
    "S",
  ],
}
`;

exports[`TrendsService getMealTrends should get meal trends for a month 1`] = `
{
  "legend": [
    "1",
    "2",
    "3",
    "4",
    "5",
  ],
  "mealTotals": {
    "breakfast": 4,
    "dinner": 3,
    "lunch": 4,
    "snack_1": 2,
    "snack_2": 1,
  },
  "mealsPerDay": [
    {
      "id": "day_1",
      "meals": [
        1,
        1,
        1,
        1,
        1,
      ],
    },
    {
      "id": "day_2",
      "meals": [
        1,
        0,
        1,
        0,
        0,
      ],
    },
    {
      "id": "day_3",
      "meals": [
        0,
        0,
        0,
        0,
        0,
      ],
    },
    {
      "id": "day_4",
      "meals": [
        1,
        1,
        1,
        0,
        1,
      ],
    },
    {
      "id": "day_5",
      "meals": [
        1,
        0,
        1,
        0,
        1,
      ],
    },
  ],
  "totalValues": 5,
}
`;

exports[`TrendsService getMealTrends should get meal trends for a week 1`] = `
{
  "legend": [
    "M",
    "T",
    "W",
    "T",
    "F",
    "S",
    "S",
  ],
  "mealTotals": {
    "breakfast": 2,
    "dinner": 1,
    "lunch": 1,
    "snack_1": 1,
    "snack_2": 0,
  },
  "mealsPerDay": [
    {
      "id": "day_M",
      "meals": [
        0,
        0,
        0,
        0,
        0,
      ],
    },
    {
      "id": "day_T",
      "meals": [
        1,
        0,
        1,
        0,
        0,
      ],
    },
    {
      "id": "day_W",
      "meals": [
        1,
        1,
        0,
        0,
        1,
      ],
    },
  ],
  "totalValues": 7,
}
`;
